by-language/javascript-typeorm/package.json

```diff
@@ -0,0 +1,22 @@
+{
+   "name": ".",
+   "version": "0.0.1",
+   "description": "Awesome project developed with TypeORM.",
+   "type": "commonjs",
+   "devDependencies": {
+      "ts-node": "10.7.0",
+      "@types/node": "^16.11.10",
+      "typescript": "4.5.2"
+   },
+   "dependencies": {
+      "typeorm": "0.3.17",
+      "reflect-metadata": "^0.1.13",
+      "pg": "^8.4.0",
+      "express": "^4.17.2",
+      "body-parser": "^1.19.1"
+   },
+   "scripts": {
+      "start": "ts-node src/index.ts",
+      "typeorm": "typeorm-ts-node-commonjs"
+   }
+}
```

Comment:
## Thoughts

CrateDB will respond with a version string like that:
```
CrateDB 5.2.2 (built a33862e/2023-02-09T10:26:09Z, Linux 5.10.124-linuxkit amd64, OpenJDK 64-Bit Server VM 19.0.2+7)
```

Not all applications/dialects/drivers like that, because they expect it to be compatible with PostgreSQL, and run feature discovery procedures on it. Does this trip TypeORM as well?

In Python lands, in order to make psycopg and asyncpg compatible with the CrateDB SQLAlchemy dialect, this was one of the details needed to be made compatible.

-- https://github.com/daq-tools/sqlalchemy-postgresql-relaxed/blob/main/sqlalchemy_postgresql_relaxed/base.py#L7-L22
-- https://github.com/crate/crate-python/pull/532

Reply:
With https://github.com/crate-workbench/typeorm/commit/fcd3841918, the example program advances to a type error, which is natural.

```
driverError: error: Cannot find data type: serial
```

Reply:
To improve the situation, https://github.com/crate-workbench/typeorm/pull/1 uses a dialect for CrateDB to provide CrateDB's [gen_random_text_uuid()](https://cratedb.com/docs/crate/reference/en/latest/general/builtins/scalar-functions.html#gen-random-text-uuid) function as an `uuidGenerator` to TypeORM.

Reply:
Within the SQL DDL statement above, the careful reader already spotted that the `NOT NULL` constraint precedes the `DEFAULT` clause, and wondered if that would work with CrateDB.

It turns out, it does not.
- https://github.com/crate/crate/issues/15156

Now that the dialect does the right thing on the `DEFAULT` definition, ...
```sql
CREATE TABLE "user" ("id" uuid NOT NULL DEFAULT gen_random_text_uuid() ...
```

... it would be so sweet if CrateDB could be supportive on that end.

Reply:
There is a patch for CrateDB already. Thanks, @seut.

- https://github.com/crate/crate/pull/15159

Reply:
In case the output of version() were still an issue, please note that since 5.1.1 it is possible to do the following:
```
cr> set search_path='public,pg_catalog';
SET OK, 0 rows affected  (0.006 sec)
cr> show search_path;
+--------------------+
| setting            |
+--------------------+
| public, pg_catalog |
+--------------------+
SHOW 1 row in set (0.007 sec)
cr> create function public.version()  RETURNS TEXT
    LANGUAGE JAVASCRIPT
    AS $$function version() {return 'test';}$$;
CREATE OK, 1 row affected  (0.057 sec)
cr> select version();
+--------+
| 'test' |
+--------+
| test   |
+--------+
SELECT 1 row in set (0.912 sec)
```

Reply:
Thank you. It will also be interesting for https://github.com/daq-tools/sqlalchemy-postgresql-relaxed.